codegen/smithy-aws-go-codegen/src/main/java/software/amazon/smithy/aws/go/codegen/RestJsonProtocolGenerator.java

```diff
@@ -44,55 +61,305 @@
+                        default:
+                            throw new CodegenException("unexpected type");
+                    }
+
+                    writer.write("return nil");
+                });
     }
 
-    @Override
-    public void generateSharedComponents(GenerationContext context) {
-        super.generateSharedComponents(context);
-        // pass
+    private void writeShapeToJsonArray(
+            Model model,
+            SymbolProvider symbolProvider,
+            GoWriter writer,
+            Shape shape
+    ) {
+        if (shape.members().size() > 1) {
+            throw new CodegenException("not possible to serialize shape with only multiple member shapes"
+                    + " to an array");
+        }
+
+        MemberShape memberShape = shape.members().iterator().next();
+        Shape targetShape = model.expectShape(memberShape.getTarget());
+
+        writer.write("array := value.Array()");
+        writer.write("defer array.Close()");
+        writer.write("");
+
+        writer.openBlock("for i := range v {", "}", () -> {
+            writer.write("av := array.Value()");
+            if (isShapeTypeDocumentSerializerRequired(targetShape.getType())) {
+                String serFunctionName = ProtocolGenerator.getDocumentSerializerFunctionName(targetShape,
+                        getProtocolName());
+                writer.openBlock("if err := $L(v[i], av); err != nil {", "}", serFunctionName, () -> {
+                    writer.write("return err");
+                });
+            } else {
+                writer.write("av" + writeSimpleShapeToJsonValue(targetShape,
+                        symbolProvider.toSymbol(targetShape), "v[i]"));
+            }
+        });
     }
 
-    @Override
-    protected void writeDefaultHeaders(GenerationContext context, OperationShape operation) {
-        super.writeDefaultHeaders(context, operation);
-        // pass
+    private void writeShapeToJsonObject(
+            Model model,
+            SymbolProvider symbolProvider,
+            GoWriter writer,
+            Shape shape,
+            Function<MemberShape, Boolean> filterMemberShapes
+    ) {
+        writer.write("object := value.Object()");
+        writer.write("defer object.Close()");
+        writer.write("");
+
+        // If the shape is a map we have special logic for enumerating members
+        if (shape.isMapShape()) {
+            writeMapShapeToJsonObject(model, symbolProvider, writer, shape);
+        } else {
+            writeShapeMembersToJsonObject(model, symbolProvider, writer, shape, filterMemberShapes);
+        }
     }
 
-    @Override
-    public void serializeInputDocument(
-            GenerationContext context,
-            OperationShape operation,
-            List<HttpBinding> documentBindings
+    private void writeShapeMembersToJsonObject(
+            Model model,
+            SymbolProvider symbolProvider,
+            GoWriter writer,
+            Shape shape,
```

Comment:
This should probably accept a `NamedMembersShape`

Reply:
I think that class might be package private.

Reply:
Yeah that class is package private, I thought about using it previously 👍